File: backend/src/main/java/com/vfms/service/ReportService.java
package com.vfms.service;

import com.vfms.dto.*;
import com.vfms.user.repository.UserRepository;
import com.vfms.fuel.repository.FuelRecordRepository;
import com.vfms.maintenance.MaintenanceRepository;
import com.vfms.trip.repository.TripRequestRepository;
import com.vfms.trip.entity.TripRequest;
import com.vfms.vehicle.Vehicle;
import com.vfms.vehicle.VehicleRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class ReportService {

    private final VehicleRepository vehicleRepository;
    private final UserRepository userRepository;
    private final TripRequestRepository tripRequestRepository;
    private final MaintenanceRepository maintenanceRepository;
    private final FuelRecordRepository fuelRecordRepository;

    public DashboardStatsDTO getDashboardStats() {
        Long totalVehicles = vehicleRepository.count();

        Map<String, Double> monthlyDistances = new LinkedHashMap<>();
        monthlyDistances.put("Jan", 4000.0);
        monthlyDistances.put("Feb", 3000.0);
        monthlyDistances.put("Mar", 2000.0);
        monthlyDistances.put("Apr", 5000.0);
        monthlyDistances.put("May", 6000.0);

        return DashboardStatsDTO.builder()
                .totalFuelCost(0.0)
                .totalMaintenanceCost(0.0)
                .totalDistance(0.0)
                .avgEfficiency(0.0)
                .totalVehicles(totalVehicles)
                .monthlyDistances(monthlyDistances)
                .build();
    }

    public CostAnalysisDTO getCostAnalysis(LocalDate startDate, LocalDate endDate) {
        return CostAnalysisDTO.builder()
                .totalMaintenanceCost(0.0)
                .totalFuelCost(0.0)
                .maintenanceTrend(new TreeMap<>())
                .fuelTrend(new TreeMap<>())
                .build();
    }

    public List<VehicleUtilizationDTO> getVehicleUtilization() {
        List<Vehicle> vehicles = vehicleRepository.findAll();
        List<TripRequest> trips = tripRequestRepository.findAll();

        return vehicles.stream().map(vehicle -> {
            List<TripRequest> vehicleTrips = trips.stream()
                    .filter(t -> t.getAssignedVehicleId() != null &&
                            t.getAssignedVehicleId().equals(vehicle.getId()))
                    .collect(Collectors.toList());

            double distance = vehicleTrips.stream()
                    .mapToDouble(t -> t.getDistanceKm() != null ? t.getDistanceKm().doubleValue() : 0.0)
                    .sum();

            return VehicleUtilizationDTO.builder()
                    .vehicleId(vehicle.getId())
                    .licensePlate(vehicle.getPlateNumber())
                    .totalDistance(distance)
                    .totalTrips((long) vehicleTrips.size())
                    .fuelConsumed(0.0)
                    .build();
        }).collect(Collectors.toList());
    }

    public List<DriverPerformanceDTO> getDriverPerformance() {
        return userRepository.findAll().stream().map(driver ->
            DriverPerformanceDTO.builder()
                    .driverId(driver.getId())
                    .driverName(driver.getFullName())
                    .totalTrips(0L)
                    .totalDistance(0.0)
                    .rating(0.0)
                    .build()
        ).collect(Collectors.toList());
    }

    public TripStatsDTO getTripStats() {
        List<TripRequest> trips = tripRequestRepository.findAll();

        long total = trips.size();
        long pending = trips.stream().filter(t -> t.getStatus() != null && "PENDING".equalsIgnoreCase(t.getStatus().name())).count();
        long approved = trips.stream().filter(t -> t.getStatus() != null && "APPROVED".equalsIgnoreCase(t.getStatus().name())).count();
        long rejected = trips.stream().filter(t -> t.getStatus() != null && "REJECTED".equalsIgnoreCase(t.getStatus().name())).count();
        long active = trips.stream().filter(t -> t.getStatus() != null && "IN_PROGRESS".equalsIgnoreCase(t.getStatus().name())).count();
        long completed = trips.stream().filter(t -> t.getStatus() != null && "COMPLETED".equalsIgnoreCase(t.getStatus().name())).count();
        long cancelled = trips.stream().filter(t -> t.getStatus() != null && "CANCELLED".equalsIgnoreCase(t.getStatus().name())).count();

        return TripStatsDTO.builder()
                .total(total)
                .pending(pending)
                .approved(approved)
                .rejected(rejected)
                .active(active)
                .completed(completed)
                .cancelled(cancelled)
                .build();
    }
}

File: frontend/src/services/reportService.ts
import api from '@/lib/api';
import { getAllFuelRecordsApi } from '@/lib/api/fuel';

export interface DashboardStats {
    totalFuelCost: number;
    totalMaintenanceCost: number;
    totalDistance: number;
    avgEfficiency: number;
    totalVehicles: number;
    monthlyDistances: Record<string, number>;
}

export interface CostAnalysis {
    totalMaintenanceCost: number;
    totalFuelCost: number;
    maintenanceTrend: Record<string, number>;
    fuelTrend: Record<string, number>;
}

export interface VehicleUtilization {
    vehicleId: number;
    licensePlate: string;
    totalDistance: number;
    totalTrips: number;
    fuelConsumed: number;
}

export interface DriverPerformance {
    driverId: number | string;
    driverName: string;
    totalTrips: number;
    totalDistance: number;
    rating: number;
}

export interface TripStats {
    total: number;
    pending: number;
    approved: number;
    rejected: number;
    active: number;
    completed: number;
    cancelled: number;
}

export interface UtilizationDepartmentData {
    name: string;
    requests: number;
}

export interface UtilizationTrendData {
    name: string;
    completed: number;
    cancelled: number;
}

export interface UtilizationSummary {
    overallFleetUtilization: number;
    departmentData: UtilizationDepartmentData[];
    completionTrend: UtilizationTrendData[];
}

type ApiEnvelope<T> = {
    success?: boolean;
    message?: string;
    data?: T;
};

function unwrapApiData<T>(payload: T | ApiEnvelope<T>): T {
    if (payload && typeof payload === 'object' && 'data' in payload) {
        return (payload as ApiEnvelope<T>).data as T;
    }

    return payload as T;
}

export const reportService = {
  getFuelLogs: async (): Promise<any[]> => {
    try {
      const { getAllFuelRecordsApi } = await import('@/lib/api/fuel');
      const result = await getAllFuelRecordsApi();
      return Array.isArray(result) ? result : [];
    } catch (error) {
      console.error('Failed to fetch fuel logs:', error);
      return [];
    }
  },

  getVehicles: async (): Promise<any[]> => {
    try {
      const { vehicleApi } = await import('@/lib/api/vehicle');
      const result = await vehicleApi.getAll();
      return Array.isArray(result) ? result : result?.data || [];
    } catch (error) {
      console.error('Failed to fetch vehicles:', error);
      return [];
    }
  },

  getRentals: async (): Promise<any[]> => {
    try {
      const { rentalApi } = await import('@/lib/api/rental');
      const result = await rentalApi.getAll();
      return Array.isArray(result) ? result : result?.data || [];
    } catch (error) {
      console.error('Failed to fetch rentals:', error);
      return [];
    }
  },

  getMaintenanceAnalytics: async (): Promise<any[]> => {
    try {
      const { maintenanceApi } = await import('@/lib/api/maintenance');
      const result = await maintenanceApi.getAll();
      return Array.isArray(result) ? result : result?.data || [];
    } catch (error) {
      console.error('Failed to fetch maintenance:', error);
      return [];
    }
  },

  async getFuelLogs(): Promise<any[]> {
    try {
      const response = await apiClient.get('/fuel');
      return response.data?.data || response.data || [];
    } catch (error) {
      console.error('Failed to fetch fuel logs:', error);
      return [];
    }
  },

  async getVehicles(): Promise<any[]> {
    try {
      const response = await apiClient.get('/vehicles');
      return response.data?.data || response.data || [];
    } catch (error) {
      console.error('Failed to fetch vehicles:', error);
      return [];
    }
  },

  async getRentals(): Promise<any[]> {
    try {
      const response = await apiClient.get('/rentals');
      return response.data?.data || response.data || [];
    } catch (error) {
      console.error('Failed to fetch rentals:', error);
      return [];
    }
  },

  async getMaintenanceAnalytics(): Promise<any[]> {
    try {
      const response = await apiClient.get('/maintenance');
      return response.data?.data || response.data || [];
    } catch (error) {
      console.error('Failed to fetch maintenance records:', error);
      return [];
    }
  },
    getDashboardStats: async (): Promise<DashboardStats> => {
        try {
            const response = await api.get<DashboardStats>('/api/reports/dashboard');
            return response.data;
        } catch (error) {
            console.error("Error fetching dashboard stats:", error);
            return {
                totalFuelCost: 0,
                totalMaintenanceCost: 0,
                totalDistance: 0,
                avgEfficiency: 0,
                totalVehicles: 0,
                monthlyDistances: {}
            };
        }
    },

    getCostAnalysis: async (startDate?: string, endDate?: string): Promise<CostAnalysis> => {
        try {
            const response = await api.get<CostAnalysis>('/api/reports/costs', {
                params: { startDate, endDate }
            });
            return response.data;
        } catch (error) {
            console.error("Error fetching cost analysis:", error);
            return {
                maintenanceTrend: {},
                fuelTrend: {},
                totalMaintenanceCost: 0,
                totalFuelCost: 0
            };
        }
    },

    getVehicleUtilization: async (): Promise<VehicleUtilization[]> => {
        try {
            const response = await api.get<VehicleUtilization[]>('/api/reports/utilization');
            return response.data;
        } catch (error) {
            console.error("Error fetching vehicle utilization:", error);
            return [];
        }
    },

    getDriverPerformance: async (): Promise<any[]> => {
        try {
            const response = await api.get<any[]>('/api/reports/driver-performance');
            return response.data;
        } catch (error) {
            console.error("Error fetching driver performance:", error);
            return [];
        }
    },

    getMaintenanceAnalytics: async () => {
        try {
            const response = await api.get<ApiEnvelope<any[]> | any[]>('/api/maintenance');
            return unwrapApiData<any[]>(response.data);
        } catch (error) {
            console.error("Error fetching maintenance analytics:", error);
            return [];
        }
    },

    getRentalAnalytics: async () => {
        try {
            const response = await api.get<ApiEnvelope<any[]> | any[]>('/api/rentals');
            return unwrapApiData<any[]>(response.data);
        } catch (error) {
            console.error("Error fetching rental analytics:", error);
            return [];
        }
    },

    getVehicles: async () => {
        try {
            const response = await api.get<ApiEnvelope<any[]> | any[]>('/api/vehicles');
            return unwrapApiData<any[]>(response.data);
        } catch (error) {
            console.error("Error fetching vehicles:", error);
            return [];
        }
    },

    getVehicleTotalCount: async () => {
        try {
            const response = await api.get<ApiEnvelope<any[]> | any[]>('/api/vehicles');
            const vehicles = unwrapApiData<any[]>(response.data);
            return vehicles.length;
        } catch (error) {
            console.error("Error fetching vehicle count:", error);
            return 0;
        }
    },

    getDriverInfractions: async () => {
        try {
            const response = await api.get<any[]>('/api/drivers/infractions');
            return response.data;
        } catch (error) {
            console.error("Error fetching driver infractions:", error);
            return [];
        }
    },

    getDriverCompliance: async () => {
        try {
            const response = await api.get<any[]>('/api/drivers/compliance');
            return response.data;
        } catch (error) {
            console.error("Error fetching driver compliance:", error);
            return [];
        }
    },

    getDriverReadiness: async () => {
        try {
            const response = await api.get<any[]>('/api/drivers/readiness');
            return response.data;
        } catch (error) {
            console.error("Error fetching driver readiness:", error);
            return [];
        }
    },

    getStaffRequests: async () => {
        return [];
    },

    getDriverLeaves: async () => {
        try {
            const response = await api.get<any[]>('/api/drivers/leaves');
            return response.data;
        } catch (error) {
            console.error("Error fetching driver leaves:", error);
            return [];
        }
    },

    getFuelLogs: async () => {
        try {
            const records = await getAllFuelRecordsApi();
            return records.map((record) => ({
                id: record.id,
                vehicleId: record.vehicleId,
                licensePlate: record.vehiclePlate,
                date: record.fuelDate,
                fuelQuantity: record.quantity,
                pricePerLiter: record.costPerLitre,
                totalCost: record.totalCost,
                odometer: record.odometerReading,
                fuelStation: record.fuelStation ?? '',
            }));
        } catch (error) {
            console.error("Error fetching fuel logs:", error);
            return [];
        }
    },

    getTripStats: async (): Promise<TripStats> => {
        try {
            const response = await api.get<TripStats>('/api/reports/trips/stats');
            return response.data;
        } catch (error) {
            console.error("Error fetching trip stats:", error);
            return {
                total: 0,
                pending: 0,
                approved: 0,
                rejected: 0,
                active: 0,
                completed: 0,
                cancelled: 0
            };
        }
    },

    getUtilizationSummary: async (): Promise<UtilizationSummary> => {
        try {
            const response = await api.get<UtilizationSummary>('/api/reports/utilization/summary');
            return response.data;
        } catch (error) {
            console.error("Error fetching utilization summary:", error);
            return {
                overallFleetUtilization: 0,
                departmentData: [],
                completionTrend: []
            };
        }
    },

    getRentals: async () => {
        try {
            const response = await api.get<ApiEnvelope<any[]> | any[]>('/api/rentals');
            return unwrapApiData<any[]>(response.data);
        } catch (error) {
            console.error("Error fetching rentals:", error);
            return [];
        }
    },

async getFuelLogs(): Promise<any[]> {
    try {
      const records = await getAllFuelRecordsApi();
      return records.map((record: any) => ({
        id: record.id,
        vehicleId: record.vehicleId,
        licensePlate: record.vehiclePlate,
        date: record.fuelDate,
        fuelQuantity: record.quantity,
        pricePerLiter: record.costPerLitre,
        totalCost: record.totalCost,
        odometer: record.odometerReading,
        fuelStation: record.fuelStation ?? "",
      }));
    } catch (error) {
      console.error('Failed to fetch fuel logs:', error);
      return [];
    }
  },

  async getVehicles(): Promise<any[]> {
    try {
      const { vehicleApi } = await import('@/lib/api/vehicle');
      const result = await vehicleApi.getAll();
      return Array.isArray(result) ? result : result?.data || [];
    } catch (error) {
      console.error('Failed to fetch vehicles:', error);
      return [];
    }
  },

  async getRentals(): Promise<any[]> {
    try {
      const { rentalApi } = await import('@/lib/api/rental');
      const result = await rentalApi.getAll();
      return Array.isArray(result) ? result : result?.data || [];
    } catch (error) {
      console.error('Failed to fetch rentals:', error);
      return [];
    }
  },

  async getMaintenanceAnalytics(): Promise<any[]> {
    try {
      const { maintenanceApi } = await import('@/lib/api/maintenance');
      const result = await maintenanceApi.getAll();
      return Array.isArray(result) ? result : result?.data || [];
    } catch (error) {
      console.error('Failed to fetch maintenance:', error);
      return [];
    }
  },
};
